src/Pulsar.Client/Internal/MessageCrypto.fs

```diff
@@ -0,0 +1,132 @@
+﻿namespace Pulsar.Client.Internal
+
+open System
+open System.Collections.Generic
+open System.IO
+open System.Security.Cryptography
+open Microsoft.Extensions.Caching.Memory
+open NSec.Cryptography
+open PemUtils
+open Pulsar.Client.Api
+open Pulsar.Client.Common
+open pulsar.proto
+open Microsoft.Extensions.Logging
+
+
+type internal MessageEncrypt(prefix: string, keyNames: string list, keyReader: ICryptoKeyReader) =
+    let prefix = prefix + " MessageEncrypt"
+    let symmetricAlgorithm = AeadAlgorithm.Aes256Gcm
+    let symmetricKey =
+        let createParameters = KeyCreationParameters(ExportPolicy = KeyExportPolicies.AllowPlaintextExport)
+        new Key(symmetricAlgorithm, &createParameters)
+    
+    let createNonce() =
+        let fixPart = ReadOnlySpan (NSec.Cryptography.RandomGenerator.Default.GenerateBytes(4))
+        Nonce(fixPart, 12 - fixPart.Length)
+    let mutable nonce = createNonce()
+
+    let parsePublicKey (rawKey: byte []) =
+        use rawStream = new MemoryStream(rawKey, writable = false)
+        use pemReader = new PemReader(rawStream)
+        pemReader.ReadRsaKey()
+    
+    let tryLoadPublicKey keyName =
+        try
+            let keyInfo = keyReader.GetPublicKey(keyName, null)
+            let rsaPublicKey = parsePublicKey(keyInfo.Key)
+            let rsa = RSA.Create()
+            rsa.ImportParameters(rsaPublicKey)
+            let encKey = rsa.Encrypt(symmetricKey.Export(KeyBlobFormat.RawSymmetricKey), RSAEncryptionPadding.OaepSHA1)
+            let encryptionKeys = EncryptionKeys(Key = keyName, Value = encKey)
+            if not (isNull keyInfo.Metadata) then
+                keyInfo.Metadata
+                |> Seq.cast<KeyValue>
+                |> Seq.iter encryptionKeys.Metadatas.Add
+            Log.Logger.LogDebug("{0} Successful load public key {1}", prefix, keyName)
+            Some encryptionKeys
+        with ex ->
+            Log.Logger.LogDebug("{0} Fail load public key {1}", prefix, keyName)
+            None
+        
+    let encryptionKeysList =
+        keyNames
+        |> List.choose tryLoadPublicKey
+    
+    member this.MayEncrypt = encryptionKeysList.Length > 0
+    
+    member this.Encrypt(msgMetadata: MessageMetadata, payload: byte []) =
+        encryptionKeysList |> List.iter msgMetadata.EncryptionKeys.Add
+        if not(Nonce.TryIncrement(&nonce)) then nonce <- createNonce()

```

Comment:
Why is nonce updated each time? What will happen if multiple threads will encrypt concurrently?

Reply:
Encryption algorithm requirement, all messages must be encrypted with a unique nonce.
Messages are encrypted in the mailbox, concurrent access to a nonce should not be.